### ISSUE TITLE
Use less ambiguous format for dates

### ISSUE THREAD

**Issue by svick** [schyntax/schyntax#8]
The `mm/dd` format is ambiguous and I think it shouldn't be used in a DSL like this.

[As far as I know](https://en.wikipedia.org/wiki/Date_format_by_country#Map), the USA is pretty much the only country that uses `mm/dd/yyyy` for dates with years, while some other countries use `dd/mm/yyyy` and others use other formats, often with different separators.

If this was about dates with years, I would suggest using [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601). But I'm not aware of any widely used standard for dates without years (and I think trimmed ISO 8601, `mm-dd`, would be also confusing).

So I'm not sure what the right solution. Maybe require spelling out the month, e.g. `Jan 1`?

**Comment by bretcope** [comment]
Some quick background: I chose `m/d`, not because it's the American format, but because it's most significant to least significant, which is how all values should be represented in my opinion. Same reason why the format which includes years is `y/m/d`, not `m/d/y` as would be the American style. I'm not sure I want to remove the slashes from the language at this point, but a better separator would be the hyphen (i.e. ISO8601).

The reason it's not the hyphen is a historical artifact of the first implementation of schyntax (called "sch" then) which used the hyphen as its range operator (`minutes(2-3)` instead of `minutes(2..3)`) and I didn't want to end up with `dates(1-2-3-4)`, which, while technically parsable, is very confusing to look at.

Now on to your points. The hyphen could be pretty easily introduced as an alternative (and preferred) separator, with the slash being deprecated. I'm in full support of this, and had been thinking about it already.

I _really_ like the idea of giving you the option to spell out the month, though. This already has precedent with the `daysOfWeek()` expression where you can specify a day either with a number or by name.

So you could write `date(2-1)` or `date(feb-1)` or `date(2016-2-1)` or `date(2016-feb-1)`.

The grammar would be something like:

```
Date
  : POSITIVE_INTEGER '-' Month '-' POSITIVE_INTEGER
  | Month '-' POSITIVE_INTEGER
  /* --- deprecated --- */
  | POSITIVE_INTEGER '/' Month '/' POSITIVE_INTEGER
  | Month '/' POSITIVE_INTEGER
;

Month
  : POSITIVE_INTEGER
  | MonthLiteral
;

MonthLiteral
  : JANUARY
  | FEBRUARY
  | MARCH
  | APRIL
  | MAY
  | JUNE
  | JULY
  | AUGUST
  | SEPTEMBER
  | OCTOBER
  | NOVEMBER
  | DECEMBER
;
```

All of the month literal terminals would be both the 3-character, and full names, case-insensitive.

```
JANUARY = \b(jan|january)\b
```

This might require moving where negative integers are parsed from the Lexer into the Parser, but that's doable.

**Comment by svick** [comment]
Yeah, assuming you're unwilling to break backwards compatibility, that sounds like a good approach to me.

**Comment by bretcope** [comment]
I'm unwilling to break it until at least a 2.0 version of the language, which is probably a ways out, if ever.

**Comment by iainhallam** [comment]
Why is there even a dates() option? There doesn't seem to be a corresponding times(). Can you just specify years(), months(), and datesOfMonth()?

**Comment by bretcope** [comment]
@iainhallam that's outside of the scope of this discussion. I'd be happy to talk to you about it, but in another issue.
